### PR TITLE
Event emitter throws an error when no "next" callback is provided

### PR DESCRIPTION
**Pull request by ms-dosx86**
## PR Type
What kind of change does this PR introduce?

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Currently it throws an error when the next callback is not provided
```ts
const emitter = new EventEmitter();
emitter.subscribe();
emitter.next();
```

## What is the new behavior?
Doesn't throw an error

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## Other information
Also moved repeated functionality into a separate function

### CODE REVIEW

File: packages/core/src/event_emitter.ts [hunk 1]
```diff
@@ -163,6 +141,17 @@ class EventEmitter_ extends Subject<any> {
 
     return sink;
   }
+
+  private createHanlder(callback: (...args: any) => void) {
```
Comment (AndrewKushnir) [inline packages/core/src/event_emitter.ts:145]: Small typo:
```suggestion
  private createHandler(callback: (...args: any) => void) {
```

File: packages/core/src/event_emitter.ts [hunk 2]
```diff
@@ -163,6 +141,17 @@ class EventEmitter_ extends Subject<any> {
...
+    if (this.__isAsync) {
+      return (...args: any) => {
+        setTimeout(() => callback(...args));
+      };
+    }
+    return (...args: any) => {
+      callback(...args);
+    };
+  }
```
Comment (AndrewKushnir) [inline packages/core/src/event_emitter.ts:154]: Can we just return callback function like this?
```suggestion
    return callback;
  }
```

File: packages/core/test/event_emitter_spec.ts [hunk 3]
```diff
@@ -30,6 +30,12 @@ import {EventEmitter} from '../src/event_emitter';
          emitter.emit(99);
        }));
 
+    it('shouldn\'t throw an error when no next callback is provided', () => {
```
Comment (AndrewKushnir) [inline packages/core/test/event_emitter_spec.ts:33]: I think the test can be updated to demonstrate a couple cases (each case might be a standalone test) when `next` callback is not present, but either `error` or `complete` callbacks are defined:
```
emitter.subscribe({
  error: () => { ... }
});
```
and
```
emitter.subscribe({
  complete: () => { ... }
});
```
and check that in both cases the corresponding callbacks are invoked.

**Review by AndrewKushnir — CHANGES_REQUESTED** [review]
Thanks for the refactoring and the fix @ms-dosx86. I've left a few comments, please have a look when you get a chance.

File: packages/core/test/event_emitter_spec.ts (hunk 3 above)
Comment (ms-dosx86) [reply packages/core/test/event_emitter_spec.ts:33]: @AndrewKushnir I think there are already tests where `error` and `complete` callbacks are tested, but a noop `next` callback is provided. I can remove `next` from those tests.

File: packages/core/src/event_emitter.ts (hunk 2 above)
Comment (ms-dosx86) [reply packages/core/src/event_emitter.ts]: Yeah, looks good

File: packages/core/test/event_emitter_spec.ts (hunk 3 above)
Comment (AndrewKushnir) [reply packages/core/test/event_emitter_spec.ts:33]: Thanks for updating the PR. I think that we should keep existing tests as is and just add a couple new ones (they'd cover different code paths).

File: packages/core/src/event_emitter.ts [hunk 4]
```diff
@@ -105,53 +105,31 @@ class EventEmitter_ extends Subject<any> {
   }
 
   subscribe(generatorOrNext?: any, error?: any, complete?: any): Subscription {
-    let schedulerFn: (t: any) => any;
+    let schedulerFn = (t: any) => {};
```
Comment (AndrewKushnir) [inline packages/core/src/event_emitter.ts:108]: I think we can also avoid creating a noop function here and just state that `schedulerFn` can actually be undefined:
```suggestion
    let schedulerFn: ((t: any) => any)|undefined;
```
The `super.subscribe` function below accepts `undefined` as an argument.

**Review by AndrewKushnir — CHANGES_REQUESTED** [review]

**Comment by AndrewKushnir** [comment]
Closing this PR due to the lack of response. Please reopen if needed. Thank you.

**Pull request closed without merging**